ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -600,6 +601,11 @@ pub trait DZKPValidator: Send + Sync {
 
     fn context(&self) -> Self::Context;
 
+    /// Sets the validator's total number of records field. This is required when using
+    /// the validate_record API, if it wasn't already set on the context used to create
```

Comment:
I was thinking that we could always set the total records for validator from the context and panic if context is not prepared correctly - do you remember why we need to support this use case (I assume we still don't expose batcher API to the protocol code, otherwise the point of this comment is moot)

Reply:
It is needed for attribution, which has a variable number of records per step. The rate of progression through records is uniform, which is required by the `validate_record` API. But the number of records per step, which for attribution is "number of users having that row number", varies.

Since it is not allowed to overwrite `total_records` with a new value once it has been set, it can't be set when creating the validator.

Reply:
ah, right - for hybrid we may get rid of it